quinn-h3/src/proto/settings.rs

```diff
@@ -0,0 +1,344 @@
+    }
+    fn encode<B: BufMut>(&self, buf: &mut B) {
+        buf.write_var(self.0);
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Settings {
+    max_header_list_size: u64,
+    qpack_max_table_capacity: u64,
+    qpack_max_blocked_streams: u64,
+}
+
+impl Default for Settings {
+    /// Create settings with H3 default values
+    fn default() -> Self {
+        Self {
+            max_header_list_size: DEFAULT_PROTO_MAX_HEADER_LIST_SIZE,
+            qpack_max_table_capacity: DEFAULT_PROTO_QPACK_MAX_TABLE_CAPACITY,
+            qpack_max_blocked_streams: DEFAULT_PROTO_QPACK_MAX_BLOCKED_STREAMS,
+        }
+    }
+}
+
+impl Settings {
+    /// Create settings with quinn-h3's recomended values
+    pub fn new() -> Self {
+        Self {
+            max_header_list_size: DEFAULT_MAX_HEADER_LIST_SIZE,
+            qpack_max_table_capacity: DEFAULT_QPACK_MAX_TABLE_CAPACITY,
+            qpack_max_blocked_streams: DEFAULT_QPACK_MAX_BLOCKED_STREAMS,
+        }
+    }
+
+    pub fn max_header_list_size(&self) -> u64 {
+        if self.max_header_list_size == 0 {
+            return std::u64::MAX;
+        }
+        self.max_header_list_size
+    }
+
+    pub fn qpack_max_table_capacity(&self) -> u64 {
+        self.qpack_max_table_capacity
+    }
+
+    pub fn qpack_max_blocked_streams(&self) -> u64 {
+        self.qpack_max_blocked_streams
+    }
+
+    pub fn set_max_header_list_size(&mut self, value: u64) -> Result<&mut Self, InvalidValue> {
+        if value > VarInt::MAX.into_inner() as u64 {
+            return Err(InvalidValue(SettingId::QPACK_MAX_TABLE_CAPACITY, value));
+        }
+        self.max_header_list_size = value;
+        Ok(self)
+    }
+
+    pub fn set_qpack_max_blocked_streams(&mut self, value: u64) -> Result<&mut Self, InvalidValue> {
+        if value > MAX_BLOCKED_STREAMS_MAX {
+            return Err(InvalidValue(SettingId::QPACK_MAX_BLOCKED_STREAMS, value));
+        }
+        self.qpack_max_blocked_streams = value;
+        Ok(self)
+    }
+
+    pub fn set_qpack_max_table_capacity(&mut self, value: u64) -> Result<&mut Self, InvalidValue> {
+        if value > MAX_TABLE_CAPACITY_MAX {
+            return Err(InvalidValue(SettingId::QPACK_MAX_TABLE_CAPACITY, value));
+        }
+        self.qpack_max_table_capacity = value;
+        Ok(self)
+    }
+
+    pub(crate) fn from_frame(settings: SettingsFrame) -> Result<Settings, Error> {
+        let mut this = Self::default();
+        for (id, val) in settings.entries[..settings.len].iter() {
+            match *id {
+                SettingId::MAX_HEADER_LIST_SIZE => this.set_max_header_list_size(*val)?,
+                SettingId::QPACK_MAX_TABLE_CAPACITY => this.set_qpack_max_table_capacity(*val)?,
+                SettingId::QPACK_MAX_BLOCKED_STREAMS => this.set_qpack_max_blocked_streams(*val)?,
+                x => return Err(Error::InvalidSettingId(x.0)),
+            };
+        }
+        Ok(this)
+    }
+
+    pub(super) fn to_frame(&self) -> SettingsFrame {
```

Comment:
Is this only for tests? If so, maybe add `#[cfg(test)]`? Otherwise, we should probably not have the `expect()` calls here? I think this would apply to everything marked `pub(super)` here.

Reply:
`to_frame()` is called in `proto::connection::with_setiings()`, and `from_frame()` in `connection::poll_recv_control()`

Reply:
So in that case we should probably make `to_frame()` fallible rather than allow it to panic?

Reply:
`to_frame()` operates on `Settings` inner-data only, so I made this function panic because if something bad happens, it can only be a programming error in the `impl` or the encapsulation is broken.

Reply:
And, if the user gets an error here, it can only be the fault of the inner Settings code. Making the caller handle that is not fair :).

Reply:
Fair enough!